types/rlp_unmarshal.go

```diff
@@ -301,86 +301,131 @@ func (t *Transaction) UnmarshalRLP(input []byte) error {
 
 // UnmarshalRLPFrom unmarshals a Transaction in RLP format
 func (t *Transaction) UnmarshalRLPFrom(p *fastrlp.Parser, v *fastrlp.Value) error {
-	elems, err := v.GetElems()
-	if err != nil {
-		return err
+	if v.Len() < 10 {
+		return fmt.Errorf("incorrect number of elements to decode transaction, expected 10 but found %d", v.Len())
 	}
 
-	if len(elems) < 9 {
-		return fmt.Errorf("incorrect number of elements to decode transaction, expected 9 but found %d", len(elems))
-	}
+	var err error
 
 	p.Hash(t.Hash[:0], v)
 
+	// Setup defaults
+	t.GasPrice = new(big.Int)
+	t.GasFeeCap = new(big.Int)
+	t.GasTipCap = new(big.Int)
+	t.Value = new(big.Int)
+	t.V = new(big.Int)
+	t.R = new(big.Int)
+	t.S = new(big.Int)
+	t.From = ZeroAddress
+
+	// Type
+	i := 0
+	if t.Type, err = ReadRlpTxType(v.Get(i)); err != nil {
```

Comment:
I feel that it is better to be explicit here and don't rely on variable which gets incremented (but provide exact index, as it was before). That way it is easier to follow code, since it must follow order from marshal.
Also it is easier to make a bug in case we add new field but forget to increment `i` some day... :slightly_smiling_face:  

WDYT?

Reply:
EDIT:
Now I realize it is not as convenient constant values, because of this if statement https://github.com/0xPolygon/polygon-edge/pull/1111/files#diff-efc5999399a6ff1082b29a52ce57fd7bfd6ebe4789d686a44c70523d2a61c684R340-R352 :thinking:

Reply:
Basically, for non-dynamic-gas-fee tx type we won't have gas fee cap and gas tip cap fields in the raw message so that's why indices could be different depending on the tx type. It won't work with the way when we explicitly specify an index until we add these two fields for any tx type, but it would make the message bigger so which would affect the network performance. Maybe not big impact, but still.

Reply:
Also, if someone forget to increment this variable, most likely some tests would fail.

Reply:
Also another point:
is it safe to introduce new fields in the middle of existing ones? It is not backward compatible (e.g. we have introduced `Type` as the first field and moved all the other existing fields by one).

Reply:
Another option is to place these two fields at the end of the message. It could work for us. Let me do that.

Reply:
Done